src/start.ts

```diff
@@ -54,265 +54,331 @@ export const handler = async (_event: any = {}): Promise<any> => {
-           */
-
-          const airnodeHDNode = ethers.utils.HDNode.fromMnemonic(
-            nodeSettings.airnodeWalletMnemonic
-          );
-          const keeperSponsorWallet = deriveKeeperSponsorWallet(
-            airnodeHDNode,
-            keeperSponsor,
-            provider
-          );
-          const requestSponsorWallet = node.evm.deriveSponsorWallet(
-            airnodeHDNode,
-            requestSponsor
-          );
-
-          // Check to prevent sending the same request for beacon update more than once
-
-          // 1. Fetch RequestedBeaconUpdate events by beaconId, sponsor and sponsorWallet
-          const requestedBeaconUpdateFilter =
-            rrpBeaconServer.filters.RequestedBeaconUpdate(
-              beaconId,
+            for (const {
+              templateId,
+              parameters,
+              oisTitle,
+              endpointName,
+              deviationPercentage,
               requestSponsor,
-              keeperSponsorWallet.address
-            );
-          const requestedBeaconUpdateEvents = await rrpBeaconServer.queryFilter(
-            requestedBeaconUpdateFilter,
-            eventLogMaxBlocks * -1
-          );
-
-          // 2. Fetch UpdatedBeacon events by beaconId
-          const updatedBeaconFilter =
-            rrpBeaconServer.filters.UpdatedBeacon(beaconId);
-          const updatedBeaconEvents = await rrpBeaconServer.queryFilter(
-            updatedBeaconFilter,
-            eventLogMaxBlocks * -1
-          );
-
-          // 3. Match these events by requestId and unmatched events
-          //    are the ones that are still waiting to be fulfilled
-          const [pendingRequestedBeaconUpdateEvent] =
-            requestedBeaconUpdateEvents.filter(
-              (rbue) =>
-                !updatedBeaconEvents.some(
-                  (ub) => rbue.args!["requestId"] === ub.args!["requestId"]
-                )
-            );
-          if (!isNil(pendingRequestedBeaconUpdateEvent)) {
-            // 4. Check if RequestedBeaconUpdate event is awaiting fulfillment by
-            //    calling AirnodeRrp.requestIsAwaitingFulfillment with requestId
-            //    and check if beacon value is fresh enough and skip if it is
-
-            const requestIsAwaitingFulfillment =
-              await airnodeRrp.requestIsAwaitingFulfillment(
-                pendingRequestedBeaconUpdateEvent.args!["requestId"]
+              eventLogMaxBlocks,
+            } of rrpBeaconServerKeeperJobs) {
+              // **************************************************************************
+              // 4. Fetch template by ID
+              // **************************************************************************
+              console.log("[DEBUG]\tfetching template...");
+              let template: [string, string, string] & {
+                airnode: string;
+                endpointId: string;
+                parameters: string;
+              };
+              try {
```

Comment:
Airnode uses a small recursivey library called [`go`](https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/utils/promise-utils.ts), which can handle retries and errors. It is named after Go's `go` action, but ironically bears little resemblance hah.

It is useful because it returns the result and error log as a tuple - so you don't need a try-catch.

This allows for logs to be returned by the function ([which Airnode also does](https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/evm/gas-prices.ts)). The return type may look like `Promise<Logs[]>`.

Logging in cloud functions slows them down, so by deferring logging you'll speed things up potentially - the downside to this is that if the function times out the logs will be lost.

None of this is critical.

Reply:
Good suggestion and I even have already thought about adding that library to Airkeeper. The thing is that I wanted to keep the code as simple as possible until the main requirements were met and the iterate over it to make it better. If you agree I'm going to create a ticket in BEC board to keep track of this improvement.

Reply:
Sounds good :)

Reply:
https://api3dao.atlassian.net/browse/BEC-84